core/src/main/java/org/elasticsearch/search/SearchModule.java

```diff
@@ -364,16 +364,59 @@ public void registerModelParser(MovAvgModel.AbstractModelParser parser) {
     }
 
     /**
-     * Enabling extending the get module by adding a custom aggregation parser.
+     * Register an aggregation.
      *
-     * @param parser The parser for the custom aggregator.
+     * @param reader reads the aggregation builder from a stream
+     * @param aggregationParser reads the aggregation builder from XContent
+     * @param aggregationName names by which the aggregation may be parsed. The first name is special because it is the name that the reader
+     *        is registered under.
      */
+    public <AB extends AggregatorBuilder<AB>> void registerAggregation(Writeable.Reader<AB> reader, Aggregator.Parser aggregationParser,
+            ParseField aggregationName) {
+        addToParserRegistry("Aggregation", aggParsers, aggregationParser, aggregationName);
+        namedWriteableRegistry.register(AggregatorBuilder.class, aggregationName.getPreferredName(), reader);
+    }
+
+    /**
+     * Register an aggregation.
+     *
+     * @deprecated prefer {@link #registerPipelineAggregation(Writeable.Reader, PipelineAggregator.Parser, ParseField)}. Will be removed
+     *             before 5.0.0GA.
+     */
+    @Deprecated
     public void registerAggregatorParser(Aggregator.Parser parser) {
-        aggParsers.add(parser);
+        addToParserRegistry("Query", aggParsers, parser, new ParseField(parser.type()));
+        namedWriteableRegistry.registerPrototype(AggregatorBuilder.class, parser.getFactoryPrototypes());
     }
 
+    /**
+     * Register a pipeline aggregation.
+     *
+     * @param reader reads the aggregation builder from a stream
+     * @param aggregationParser reads the aggregation builder from XContent
+     * @param aggregationName names by which the aggregation may be parsed. The first name is special because it is the name that the reader
+     *        is registered under.
+     */
+    public <AB extends PipelineAggregatorBuilder<AB>> void registerPipelineAggregation(Writeable.Reader<AB> reader,
```

Comment:
I don't know what it looks like in query registration, but in all the other register methods we have, we put the "key" first. ie here the agg name should be first? I really don't understand why we are taking ParseField instead of String too...seems we will never get rid of camelCase alternative fields if we keep extending support for it.